Wrappers/Python/cil/optimisation/algorithms/Algorithm.py

```diff
@@ -31,167 +87,140 @@ class Algorithm(object):
                                        and so forth. This is by default 1 and should be increased\
                                        when evaluating the objective is computationally expensive.
         :type update_objective_interval: int, optional, default 1
-        :param log_file: log verbose output to file
-        :type log_file: str, optional, default None
         '''
         self.iteration = -1
-        self.__max_iteration = kwargs.get('max_iteration', 0)
+        self.__max_iteration = max_iteration
         self.__loss = []
         self.memopt = False
         self.configured = False
-        self.timing = []
         self._iteration = []
-        self.update_objective_interval = kwargs.get('update_objective_interval', 1)
+        self.update_objective_interval = update_objective_interval
         # self.x = None
         self.iter_string = 'Iter'
-        self.logger = None
-        self.__set_up_logger(kwargs.get('log_file', None))
-        
+        if 'log_file' in kwargs:
+            warn("use `run(callbacks=[LogfileCallback(log_file)])` instead of `log_file`",
+                 DeprecationWarning, stacklevel=2)
+            self.__log_file = kwargs['log_file']
+
     def set_up(self, *args, **kwargs):
         '''Set up the algorithm'''
-        raise NotImplementedError()
+        raise NotImplementedError
     def update(self):
         '''A single iteration of the algorithm'''
-        raise NotImplementedError()
-    
+        raise NotImplementedError
+
     def should_stop(self):
         '''default stopping criterion: number of iterations
-        
+
         The user can change this in concrete implementation of iterative algorithms.'''
         return self.max_iteration_stop_criterion()
-    
-    def __set_up_logger(self, fname):
-        """Set up the logger if desired"""
-        if fname:
-            print("Will output results to: " +  fname)
-            handler = logging.FileHandler(fname)
-            self.logger = logging.getLogger("obj_fn")
-            self.logger.setLevel(logging.INFO)
-            self.logger.addHandler(handler)
-    
+
+    def __set_up_logger(self, *_, **__):
+        warn("use `run(callbacks=[LogfileCallback(log_file)])` instead", DeprecationWarning, stacklevel=2)
+
     def max_iteration_stop_criterion(self):
         '''default stop criterion for iterative algorithm: max_iteration reached'''
         return self.iteration > self.max_iteration
 
     def __iter__(self):
         '''Algorithm is an iterable'''
         return self
-    def next(self):
-        '''Algorithm is an iterable
-        
-        python2 backwards compatibility'''
-        return self.__next__()
+
     def __next__(self):
         '''Algorithm is an iterable
-        
-        calling this method triggers update and update_objective
+
+        This method triggers :code:`update()` and :code:`update_objective()`
```

Comment:
I wonder if triggering `update_objective` shouldn't be part of the new callback rather than `__next__`

Reply:
surely `update_objective` is integral to the concept of `Algorithm` though? i.e. callbacks by comparison are optional so `callbacks=[UpdateObjectiveCallback()]` would imply users may want to *omit* `UpdateObjectiveCallback()`.